### PR TITLE
[GraphOptz] EliminateSliceConcat handles reshape

### LINKED ISSUE #4628: [GraphOptz] Improve EliminateSliceConcat optimization
Stumbled over this when compiling some GRU network:
image
So the slicing is done over dimension 0 (resulting in singular dimensions) and the concatenation along dimension 1 but in this particular case all the Slice and Concat node should be eliminated and replaced with a Reshape.

### PR DESCRIPTION
**Pull request by Squadrick**
Summary: If the slice dimension is not the same as concat dimension, it may be possible to replace the slices with a merged slice + reshape if the slice dimension is adjacent to the concat dimension.

Documentation: Comments added in code to explain new optimization.

Fixes #4628 

Test Plan: A new test case is added to check for this optimization: `EliminateSliceConcatWithReshapeTest`.

### CODE REVIEW

**Comment by Squadrick** [comment]
cc @jfix71 for review

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -2523,59 +2549,107 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
...
+    //
+    // Store consecutive slices along *any* dimension. If the consecutive
+    // slices' dimension is the same as the concat, the concat can be removed.
+    // If the slices' dimension is different from the concat, the nodes
+    // can be replaced with a single slice+reshape.
+    std::vector<std::pair<unsigned_t /* dimension of slicing */,
+                          std::vector<SliceNode *>>> consecutiveSlices;
     std::vector<SliceNode *> currConsecutiveSlices;
     SliceNode *lastSN = nullptr;
+    unsigned_t lastDim = -1;
```
Comment (Squadrick) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2561]: Since it's unsigned, the default value will be `2**32-1`. I hope that's not a valid dimension I would have to consider.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -2523,59 +2549,107 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
...
 
     // Mapping from old Slices to new Slices where the range of each old Slice
     // is a subset of the corresponding new Slice
     std::unordered_map<SliceNode *, SliceNode *> oldToNewSlices;
-    for (const auto &slices : consecutiveSlices) {
+
+    // Mapping from new fused slices to reshape nodes that maybe be needed in
+    // case the fused slices need to be reshaped before concat. This happens
+    // when consecutive slices dimension differs from concat dimension by 1.
+    std::unordered_map<SliceNode *, ReshapeNode *> newSlicesToReshape;
```
Comment (Squadrick) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2596]: Just thought of an alternate approach to create a single mapping as such:
```c++
std::unordered_map<SliceNode *, NodeValue *> oldToNewSlicesOrReshape;
```
and similarly, change the lookup logic. This would have a bit more casting but will use less memory (single hashmap vs. two). 

If warranted, I'll upload the new patch to have that change.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 3]
```diff
@@ -2523,59 +2549,107 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
...
         oldToNewSlices[slice] = newSlice;
       }
+
+      // Create a reshape node based on consecutive slice dimension and
+      // concat dimension.
+      if (slicesDim == CN->getDim() + 1 ||
+          slicesDim == CN->getDim() - 1) {
+        auto outputDimVec = newSlice->getResult().dims().vec();
+        outputDimVec[CN->getDim()] *= outputDimVec[slicesDim];
+        outputDimVec[slicesDim] = 1;
```
Comment (Squadrick) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2639]: Thought about this logic, and did a couple of experiment in Numpy, and it seems to hold. Still 100% sure, so some extra attention during the review may be warranted.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 2 above)
Comment (Squadrick) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp]: Done. Much cleaner code. It uses `std::unordered_map<SliceNode *, Node *>`.

File: tests/unittests/GraphOptzTest.cpp [hunk 4]
```diff
@@ -5451,6 +5451,31 @@ TEST_F(GraphOptz, EliminateSliceConcatTest) {
   checkNumericalEquivalence();
 }
 
+TEST_F(GraphOptz, EliminateSliceConcatWithReshapeTest) {
```
Comment (jfix71) [inline tests/unittests/GraphOptzTest.cpp:5454]: Can you add an explanation of the test, e.g. perhaps with ASCII art? 🙂

Comment (jfix71) [reply tests/unittests/GraphOptzTest.cpp:5454]: Can we also add another test a little more complex, e.g. with more inputs into the Concat other than just the three consecutive slices? Better yet, a Concat on dim1, with 3 consecutive slices on dim0 and another 3 consecutive slices on dim2.

File: tests/unittests/GraphOptzTest.cpp [hunk 5]
```diff
@@ -5451,6 +5451,31 @@ TEST_F(GraphOptz, EliminateSliceConcatTest) {
...
+
+  // src, A, B, C, CN.
+  EXPECT_EQ(F_->getNodes().size(), 5);
+
+  optimizedF_ = optimizeFunction(
+      F_, {FunctionPassID::EliminateSliceConcat, getDCEPassConfig()});
+
+  // src, FUSED_SLICE[A, B, C], RESHAPE, CN.
+  // FUSED_SLICE, RESHAPE are inserted by optimization.
+  EXPECT_EQ(optimizedF_->getNodes().size(), 4);
```
Comment (jfix71) [inline tests/unittests/GraphOptzTest.cpp:5472]: Can you actually verify the structure of the graph instead of just putting what the expected nodes are after the opt? E.g. that all of the dimensions, concat axis (dim), etc. make sense.

File: tests/unittests/GraphOptzTest.cpp [hunk 6]
```diff
@@ -5451,6 +5451,31 @@ TEST_F(GraphOptz, EliminateSliceConcatTest) {
...
+
+  bindings_.allocate(src)->getHandle<float>().randomize(-10.0, 10.0,
+                                                        mod_.getPRNG());
+  checkNumericalEquivalence();
```
Comment (jfix71) [inline tests/unittests/GraphOptzTest.cpp:5476]: You can specify the allowed error here -- let's make it 0.f since there's only data movement in this op.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 7]
```diff
@@ -2523,70 +2550,114 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
...
+    // Store consecutive slices along *any* dimension. If the consecutive
+    // slices' dimension is the same as the concat, the concat can be removed.
+    // If the slices' dimension is different from the concat, the nodes
+    // can be replaced with a single slice+reshape.
+    std::vector<std::pair<unsigned_t /* dimension of slicing */,
+                          std::vector<SliceNode *>>>
+        consecutiveSlices;
     std::vector<SliceNode *> currConsecutiveSlices;
     SliceNode *lastSN = nullptr;
+    unsigned_t lastDim = -1;
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2563]: Do you encounter compilation issues if you use int instead of unsigned?

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 3 above)
Comment (jfix71) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2632]: Yeah, I think this is wrong. As an example, in your unit test, change the first few lines to the following, and it will fail:

```cpp
  auto *src =
      mod_.createPlaceholder(ElemKind::FloatTy, {8, 5, 3}, "src", false);
  auto *A = F_->createSlice("A", src, {0, 0, 0}, {2, 5, 3});
  auto *B = F_->createSlice("B", src, {2, 0, 0}, {4, 5, 3});
  auto *C = F_->createSlice("C", src, {4, 0, 0}, {6, 5, 3});
```

I think you want to use the original dimension of one of the slices.

File: tests/unittests/GraphOptzTest.cpp [hunk 8]
```diff
@@ -5451,6 +5451,31 @@ TEST_F(GraphOptz, EliminateSliceConcatTest) {
   checkNumericalEquivalence();
 }
 
+TEST_F(GraphOptz, EliminateSliceConcatWithReshapeTest) {
+  auto *src =
+      mod_.createPlaceholder(ElemKind::FloatTy, {4, 5, 3}, "src", false);
+  auto *A = F_->createSlice("A", src, {0, 0, 0}, {1, 5, 3});
+  auto *B = F_->createSlice("B", src, {1, 0, 0}, {2, 5, 3});
+  auto *C = F_->createSlice("C", src, {2, 0, 0}, {3, 5, 3});
```
Comment (jfix71) [inline tests/unittests/GraphOptzTest.cpp:5459]: As mentioned above:
```suggestion
  auto *src =
      mod_.createPlaceholder(ElemKind::FloatTy, {8, 5, 3}, "src", false);
  auto *A = F_->createSlice("A", src, {0, 0, 0}, {2, 5, 3});
  auto *B = F_->createSlice("B", src, {2, 0, 0}, {4, 5, 3});
  auto *C = F_->createSlice("C", src, {4, 0, 0}, {6, 5, 3});
```

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 3 above)
Comment (Squadrick) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2632]: The optimization will only work if the size of the slice along the consecutive dimension is 1. This did bring up another problem, that I was not checking if this was the case. A check, along with a comment, has been added. So in your example, this optimization will do nothing.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 7 above)
Comment (Squadrick) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2563]: Switched to `int`.

File: tests/unittests/GraphOptzTest.cpp (hunk 8 above)
Comment (Squadrick) [reply tests/unittests/GraphOptzTest.cpp:5459]: Explained in a previous comment. Only works if the slice size along the consecutive dimension is 1.

**Comment by Squadrick** [comment]
@jfix71 Thanks for the review. Made the changes you requested. 

There is one significant change to the logic. If the slice consecutive dimension is after the concat dimension, there's a need for a transpose node: merged slice + transpose + reshape. Added the logic, and updated the comments. Could you take a look at that? 

The remaining work is around making the test better: ASIC art + check for structure. I'll push some commits soon for this.

Also, during graph optimization, what is the heuristic to optimize for? Is the sole focus to minimize the number of nodes, or should we take the type of node into consideration? Is it safe to assume that 4 slices should always be optimized to 1 slice + 1 transpose + 1 reshape?

**Comment by Squadrick** [comment]
@jfix71 Did you get any time to take another look at this?

**Comment by jfix71** [comment]
Really sorry for the major delays here, I've been super swamped. I will definitely get to reviewing next week.

**Comment by Squadrick** [comment]
@jfix71 No problem, take your time. Just didn't want the bot to auto-close for being stale for too long :).

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 9]
```diff
@@ -2536,70 +2551,135 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
...
+      if (slicesDim != CN->getDim() &&
+          ((slicesDim != CN->getDim() + 1 && slicesDim != CN->getDim() - 1) ||
+           slices[0]->getResult().dims()[slicesDim] != 1)) {
+        // Optimizations are possible only if:
+        // 1) slices consecutive dimension is the same as concat dimension, or
+        // 2) slices consecutive dimension is adjacent to the concat dimension,
+        //    and the size of each slice along the consecutive dimension is 1.
+        //    NOTE: Checking the size of 0th slice is sufficient, as opposed
+        //    to checking every slice. If the slices can be concatenated,
+        //    each slice size must be equal.
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2616]: Just to make this more clear, can we reword it to something like:

NOTE: Checking the slicesDim dimension of 0th slice is sufficient, as opposed to checking every slice. If the slices can be concatenated (and they're being concatenated along a different dimension), then each slicesDim dim must be equal.

File: tests/unittests/GraphOptzTest.cpp [hunk 10]
```diff
@@ -5525,6 +5525,38 @@ TEST_F(GraphOptz, EliminateSliceConcatTest) {
...
+
+  // src, A, B, C, CN.
+  EXPECT_EQ(F_->getNodes().size(), 11);
+
+  optimizedF_ = optimizeFunction(
+      F_, {FunctionPassID::EliminateSliceConcat, getDCEPassConfig()});
+
+  // src, FUSED_SLICE[A, B, C], RESHAPE, CN.
+  // FUSED_SLICE, RESHAPE are inserted by optimization.
+  EXPECT_EQ(optimizedF_->getNodes().size(), 9);
```
Comment (jfix71) [inline tests/unittests/GraphOptzTest.cpp:5553]: Can you do some actual testing of how many of each node is leftover? Using `countNodeKind()`, as seen in other tests.

**Review by jfix71 — COMMENTED** [review]
Hey @Squadrick, apologies again for all the delays. I think this is pretty close to landing -- one request would be to add in a test for any edge cases that the optimization should not apply for (e.g. the example I had specified in my comment, where the adjacent dims are not 1). We've had some opts in the past for slicing/concatenating where we ended up with numerical issues, and I'd like to make sure we have good coverage for cases both where the opt will and won't apply.

File: tests/unittests/GraphOptzTest.cpp (hunk 10 above)
Comment (Squadrick) [reply tests/unittests/GraphOptzTest.cpp:5709]: Done.

**Review by Squadrick — COMMENTED** [review]
@jfix71 I've added another test: `EliminateSliceConcatWithReshapeTestNoChange` that ensures no optimization occurs when slice and concat dims are not adjacent. I've also made the changes you requested, including using `countNodeKind`.

**Review by jfix71 — APPROVED** [review]
LGTM, thanks for this contribution and your patience @Squadrick!
